Fix RTL alignment / VerticalDirection bug on Layout Explorer

adalberht opened the pull request:
This CL should fixes #2138 

@jacob314

jacob314 commented on packages/devtools_app/lib/src/inspector/inspector_data_models.dart:
```diff
@@ -370,6 +370,9 @@ class FlexLayoutProperties extends LayoutProperties {
   final TextBaseline textBaseline;
 
+  @override
+  List<LayoutProperties> get children => flipMainAxis ? super.children.reversed.toList() : super.children;
```
maybe leave children in the real order and add a second getter called 
displayChildren that are in the display order? You should be able to rename children n the parent class to displayChildren and then add a new children getter to the parent class to make this change without having to do much manual work.

jacob314 commented on packages/devtools_app/lib/src/inspector/inspector_data_models.dart:
```diff
@@ -417,6 +420,36 @@ class FlexLayoutProperties extends LayoutProperties {
+  bool get flipMainAxis {
+    if (_flipMainAxis != null) return _flipMainAxis;
+    assert(direction != null);
+    final startIsTopLeft = () {
```
make this a method instead of a closure.
e.g
`bool startIsTopLeft() {`

adalberht commented:
Hey Jacob, sorry for my inactivity on this PR. I was busy with getting involved on hosting the Flutter Indonesia hackathon for the past month. I just got the time to work on this PR again but I need more time on fixing and writing the proper tests.
I'll let you know once this PR is ready for review again!

adalberht commented on packages/devtools_app/test/layout_explorer/layout_explorer_serialization_delegate.dart:
```diff
@@ -0,0 +1,69 @@
+
+/// This class is merely a duplicate of the InspectorSerializationDelegate inside the `assets/scripts/inspector_polyfill_script.dart`.
+/// This delegate is used for testing Widget Diagnostics Node so that we don't have to create manual JSON each time we want to create new test cases.
+/// TODO(adalberht): Ask Jacob on what's the better solution on code reuse between the Layout Explorer polyfillscripts and the code inside test package?
```
I wanted to reuse some code on the polyfill scripts but I'm not sure on how to do that properly without cloning/copy-pasting them.
This is useful so that our unit tests can just generate the Layout Explorer Remote Diagnostics Node based on actual widgets instead of us creating JSON data for each tests cases that we want to create.
But let me know if that's not preferrable! To be honest writing it this way is way makes writing the tests more concise and pleasant :D

jacob314 replied:
Clever idea! Lets land like this and we can add a build script to copy the code so it can be kept completely in sync. The other way we could solve this is by tests communicating with a remote Dart vm like in the inspector_service tests. That has the downside that the tests are slower to run.

adalberht replied:
Awesome, thanks, Jacob! Alright then, I agree maybe let's land like this and l will try to look on the alternative testing methods with remote Dart VM that you mentioned

adalberht commented on packages/devtools_app/test/layout_explorer/layout_explorer_test_utils.dart:
```diff
@@ -0,0 +1,30 @@
+
+import 'layout_explorer_serialization_delegate.dart';
+
+Future<RemoteDiagnosticsNode> widgetToLayoutExplorerRemoteDiagnosticsNode({
+  @required Widget widget,
+  @required WidgetTester tester,
+  int subtreeDepth = 1,
+}) async {
+  await tester.pumpWidget(
+    MaterialApp(home: Scaffold(body: widget))
+  );
+  final element = find.byWidget(widget).evaluate().first;
+  final nodeJson =
+  element.toDiagnosticsNode(style: DiagnosticsTreeStyle.dense).toJsonMap(
+    LayoutExplorerSerializationDelegate(
+      subtreeDepth: subtreeDepth,
+      service: WidgetInspectorService.instance,
+    ),
+  );
+  return RemoteDiagnosticsNode(nodeJson, null, null, null);
+}
```
This is what I meant by my previous comment

jacob314 replied:
I like it!

jacob314 commented:
You will want to remove the draft tag and fix the couple of lint errors causing the bots to fail. 
For the use of a protected member error, you can use the quick fix to allow that lint error for that line.

adalberht commented:
Hey Jacob, sorry that I missed out on some test cases on my previous version of the PR, apparently we also need to flip the main axis alignment because that affects the calculation on the free spaces.
Can you please kindly take a look at it again? I added some lines on the main axis and free spaces calculation and also added some more unit tests for it.

adalberht commented:
Awesome, thanks @jacob314 !
I didn't have write access to the repository, so you can merge it whenever you want! :D
